[10.x] Add a `File::bytesToHuman()` helper

emmadesilva opened the pull request:
## Abstract

Adds a `File::bytesToHuman()` helper to format the number of bytes to a human-readable string.

### Functionality

The helper works by iterating through the input byte count and dividing it by `1024`. The iteration count is the used in a lookup array to get the appropriate byte suffix. It supports rounding to a specific precision. Results are also slightly rounded up.

### Development

I was originally thinking this could be a helper in the `Str` class, but after bringing it up in `#internals` we thought it might be better in the `File` facade. I am of course more than happy to move it to the `Str` class. **Thanks to all the great people in the Discord!**

### Added tests

I added a test with 8 assertions to get a broad range of conditions. I also ran some benchmarks on different implementations, and all of them have minuscule performance concerns (~0.00058ms avg/1 million iterations).

### Backwards compatibility 

This adds a new feature, so it will not break any existing features. It can thus also go in a minor release version.

### How this improves Laravel

This is one of those things I often add to Laravel applications as a `Str` macro. And I thought it is so useful, so I wanted to share it with everyone!

emmadesilva committed:
Add a `File::bytesToHuman()` helper

joshuaruesweg commented on src/Illuminate/Filesystem/Filesystem.php:
```diff
@@ -789,4 +789,20 @@ public function cleanDirectory($directory)
+            $bytes /= 1024;
+        }
+
+        return round($bytes, $precision).' '.['B', 'KB', 'MB', 'GB', 'TB'][$i];
```
That will not work for Petabyte: 

https://3v4l.org/aUKD7

So please use a fallback if we have more then the highest unit.

joshuaruesweg replied:
Also: Maybe use `sprintf` for text output. This should imo be more readable.

caendesilva replied:
Sure thing! I could increase the lookup array to Yottabytes? I can of course also add a fallback, or both. What do you think is best?

Will use `sprintf` as well.

joshuaruesweg replied:
Personally, I would like to see it expanded to yottabytes when it is irrelevant whether it is needed or not, it is there and the function can then. 

Additionally, you always need a fallback, as there can always be sizes you don't care about in the code (even if the integer is limited, you wan't always have a fallback).

So the easiest way for a fallback is a break within the loop, if we reach a certain point (or use a `for` loop with a break if `$bytes < 1024`).

caendesilva replied:
Thanks for the feedback!

7d046255e7ccd9f965c9bbc07345ccc8e8f907a2 adds a cap so that the iterator stops dividing when reaching the end of the units array which now also supports extreme values. Also uses `sprintf` as requested. Also adds tests for large numbers.

Let me know if you want me to squash the commits!

emmadesilva committed:
Improve the `File::bytesToHuman()` helper implementation

Adds a cap so that the iterator stops dividing when reaching the end of the units array. Also uses `sprintf` as requested. Also adds tests for large numbers.

emmadesilva committed:
Fix style

taylorotwell commented:
Hey @caendesilva - I started digging into what Rails does here.

https://api.rubyonrails.org/classes/ActionView/Helpers/NumberHelper.html

This may make sense as part of a new `Number` utility class like `Str`? Maybe we should pursue that instead?

caendesilva commented:
> Hey @caendesilva - I started digging into what Rails does here.
> 
> https://api.rubyonrails.org/classes/ActionView/Helpers/NumberHelper.html
> 
> This may make sense as part of a new `Number` utility class like `Str`? Maybe we should pursue that instead?

@taylorotwell That sounds even better to me! I'd love to help. I'm in the `#internals` Discord if you want to chat! (username `@Caen`)

amorim778 commented:
how to install this update?

caendesilva commented:
> how to install this update?

A version of it will be released tomorrow! https://laravel-news.com/laravel-number-utility-class

badasukerubin commented:
I assume that there could be a reverse `sizeToBytes` or `humanToBytes` method.
